### DISCUSSION TITLE
time get input port for custom BlockDriver/TimeManager component

### DISCUSSION THREAD

**Discussion by beavuhs** [nasa/fprime#2712] [Q&A] [answered]
I am currently building a custom block driver/time setter component that drives rate groups based on a clock from an external simulation. This component also needs to read in time from that simulation and set the FSW's clock accordingly. I have the first part figured out, but am struggling to hook up this component to the `time connections` in my topology. The error I am getting says `instance <myCompnent> has no time get input port`. I am defining a `time get port` in the fpp, and I have directly copied the function defintion of `timeGetPort_handler` from the LinuxTimeImpl.cpp file into my component. What else am I missing here to be able to use this component as my time source?

**Comment by LeStarch** [comment, marked as answer]
Does the definition of the time get port on `myComponent`'s model look exactly like the following?

```
sync input port timeGetPort: Fw.Time
```

**Reply by LeStarch** [reply]
Note: the following syntax is for consumers of time.  The above syntax is for suppliers of time, and is what you need on your custom tick/time component.

```
@ This is for consumers (users) of the time system, not suppliers.
time get port timeCaller
```

**Reply by beavuhs** [reply]
That worked! Thank you.

**Reply by LeStarch** [reply]
Happy to help!  Good luck!

**Reply by beavuhs** [reply]
Follow up question @LeStarch, it appears that our sim is running too quickly for the FSW to keep up. I am seeing several `Rate group slipped on cycle X` in the events tab of the GDS. Sometimes, the FSW crashes with the following assert: `Assert in file build-fprime-automatic-native/F-Prime/Svc/ActiveRateGroup/ActiveRateGroupComponentAc.cpp, line 770: 8`. Are you aware of what might be causing this ASSERT? 
![image](https://github.com/nasa/fprime/assets/113480361/7bb46539-b8eb-45e7-871c-56e0ec072628)

**Reply by LeStarch** [reply]
Yes.  Asserts of the form `*Ac.cpp:  ...  8` are "queue overflow".  You have dispatched enough messages to have overfilled the queue before the component could process the messages.

While you could increase queue size, I tend to agree with you: the simulation is ticking the rate-group too fast and the flight software is not keeping up.

Can I ask how fast the rate-group is intended to run and how fast the simulation is running it?  Is this running in Linux, some sort of containerized environment, or something else?

**Reply by beavuhs** [reply]
The rate group is meant to run at 25 Hz, and the simulation can be running up to 100x faster than real time.  Yes, this is running in a containerized environment.

**Reply by LeStarch** [reply]
My guess is that the thread dispatching inside a container is too slow to keep up with `2500Hz`.  This may be compounded if the containerization is also emulating (e.g. QEMU).

I might suggest this: clock your  simulation at `1x` and then ramp up the speed-multiplier? That will show you where it starts to break down.  You can also try to reduce the overhead of the container/emulation if possible....but knowing what multiplier starts to break will tell you "how much" overhead must be dealt with.

**Reply by timcanham** [reply]
The other thing you can do is to add a component to the end of the rate group to start the next cycle. That way the simulation can auto-scale depending on the speed of the container.

**Reply by beavuhs** [reply]
@LeStarch I should maybe clarify: both the sim and the FSW are running in a container on a Linux VM.

**Reply by LeStarch** [reply]
My point is that containers often have overhead and as such can be slower than traditional systems.  Hence looking at specifically at how fast things can be clocked.

**Reply by beavuhs** [reply]
This works beautifully, thanks a ton. Able to run sims at 100x realtime with FSW in the loop.

**Reply by ephraim71** [reply]
@Jmb5501 For some reason, I still get system time when using this->getTime();
Is that what you used to get current time from the component?

**Reply by LeStarch** [reply]
`this->getTime()` will always return time from the Svc/Interfaces/TimeInterface implementation.  This defaults to PosixTime or C++ ChronoTime which both use system time.  If you need a different time, you should implement your own time handler.

**Reply by ephraim71** [reply]
@LeStarch Thanks this worked!

**Reply by ephraim71** [reply]
@timcanham @Jmb5501 - A question on this : 
`The other thing you can do is to add a component to the end of the rate group to start the next cycle. That way the simulation can auto-scale depending on the speed of the container.`
A bit confused on how to implement this. Lets say 3 active components are connected to a rate group, and the 4th component is the component which will start the next cycle. 
Now, don't active components run in parallel? Is it possible that the 4th component starts executing before the first 3 are complete?
